modules/queue/manager.go

```diff
@@ -0,0 +1,266 @@
+		Pool:          pool,
+	}
+	m.mutex.Lock()
+	m.counter++
+	desc.QID = m.counter
+	desc.Name = fmt.Sprintf("queue-%d", desc.QID)
+	if named, ok := queue.(Named); ok {
+		desc.Name = named.Name()
+	}
+	m.Queues[desc.QID] = desc
+	m.mutex.Unlock()
+	log.Trace("Queue Manager registered: %s (QID: %d)", desc.Name, desc.QID)
+	return desc.QID
+}
+
+// Remove a queue from the Manager
+func (m *Manager) Remove(qid int64) {
+	m.mutex.Lock()
+	delete(m.Queues, qid)
+	m.mutex.Unlock()
+	log.Trace("Queue Manager removed: QID: %d", qid)
+
+}
+
+// GetDescription by qid
+func (m *Manager) GetDescription(qid int64) *Description {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+	return m.Queues[qid]
+}
+
+// Descriptions returns the queue descriptions
+func (m *Manager) Descriptions() []*Description {
+	m.mutex.Lock()
+	descs := make([]*Description, 0, len(m.Queues))
+	for _, desc := range m.Queues {
+		descs = append(descs, desc)
+	}
+	m.mutex.Unlock()
+	sort.Sort(DescriptionList(descs))
+	return descs
+}
+
+// Workers returns the poolworkers
+func (q *Description) Workers() []*PoolWorkers {
+	q.mutex.Lock()
+	workers := make([]*PoolWorkers, 0, len(q.PoolWorkers))
+	for _, worker := range q.PoolWorkers {
+		workers = append(workers, worker)
+	}
+	q.mutex.Unlock()
+	sort.Sort(PoolWorkersList(workers))
+	return workers
+}
+
+// RegisterWorkers registers workers to this queue
+func (q *Description) RegisterWorkers(number int, start time.Time, hasTimeout bool, timeout time.Time, cancel context.CancelFunc) int64 {
+	q.mutex.Lock()
+	defer q.mutex.Unlock()
+	q.counter++
+	q.PoolWorkers[q.counter] = &PoolWorkers{
+		PID:        q.counter,
+		Workers:    number,
+		Start:      start,
+		Timeout:    timeout,
+		HasTimeout: hasTimeout,
+		Cancel:     cancel,
+	}
+	return q.counter
+}
+
+// CancelWorkers cancels pooled workers with pid
+func (q *Description) CancelWorkers(pid int64) {
+	q.mutex.Lock()
+	pw, ok := q.PoolWorkers[pid]
+	q.mutex.Unlock()
+	if !ok {
+		return
+	}
+	pw.Cancel()
+}
+
+// RemoveWorkers deletes pooled workers with pid
+func (q *Description) RemoveWorkers(pid int64) {
+	q.mutex.Lock()
+	delete(q.PoolWorkers, pid)
+	q.mutex.Unlock()
+}
+
+// AddWorkers adds workers to the queue if it has registered an add worker function
+func (q *Description) AddWorkers(number int, timeout time.Duration) context.CancelFunc {
+	if q.Pool != nil {
+		// the cancel will be added to the pool workers description above
+		return q.Pool.AddWorkers(number, timeout)
+	}
+	return nil
+}
+
+// NumberOfWorkers returns the number of workers in the queue
+func (q *Description) NumberOfWorkers() int {
+	if q.Pool != nil {
+		return q.Pool.NumberOfWorkers()
+	}
+	return -1
+}
+
+// MaxNumberOfWorkers returns the maximum number of workers for the pool
+func (q *Description) MaxNumberOfWorkers() int {
+	if q.Pool != nil {
+		return q.Pool.MaxNumberOfWorkers()
+	}
+	return 0
+}
+
+// BoostWorkers returns the number of workers for a boost
+func (q *Description) BoostWorkers() int {
```

Comment:
Why are workers added in batches that need to be removed in block? Seems like a linked list should work better and be more granular; unless of course workers are needed in blocks specifically.

Reply:
You would need a context and channel for every worker which is quite inefficient